### ISSUE TITLE
Incorrect handling of arithmetic with folded datetimes

### ISSUE THREAD

**Issue by bdkearns** [dateutil/dateutil#344]
Adding a timedelta to a folded datetime produces incorrect result, see script/output. Works fine with equivalent pytz timezone.
```
import datetime
import dateutil.tz
#import pytz

d1 = datetime.datetime(2016, 11, 6, 6, 12, 15, 0, datetime.timezone.utc)
print(repr(d1))
print(d1.timestamp())
d2 = d1.astimezone(dateutil.tz.gettz('US/Eastern'))
#d2 = d1.astimezone(pytz.timezone('US/Eastern'))
print(repr(d2))
print(d2.timestamp())
d3 = d2 + datetime.timedelta(microseconds=0)
print(repr(d3))
print(d3.timestamp())

assert d1 == d2 == d3
assert d1.timestamp() == d2.timestamp() == d3.timestamp()
```

```
$ python dateutil_bug.py
datetime.datetime(2016, 11, 6, 6, 12, 15, tzinfo=datetime.timezone.utc)
1478412735.0
_DatetimeWithFold(2016, 11, 6, 1, 12, 15, tzinfo=tzfile('/usr/share/zoneinfo/US/Eastern'))
1478412735.0
datetime.datetime(2016, 11, 6, 1, 12, 15, tzinfo=tzfile('/usr/share/zoneinfo/US/Eastern'))
1478409135.0
Traceback (most recent call last):
  File "dateutil_bug.py", line 17, in <module>
    assert d1.timestamp() == d2.timestamp() == d3.timestamp()
AssertionError
```

```
$ python
Python 3.5.2 (default, Nov 17 2016, 17:05:23)
[GCC 5.4.0 20160609] on linux
```

**Comment by pganssle** [comment]
I'm fairly certain `dateutil`'s behavior is the correct behavior, after [PEP 495](https://www.python.org/dev/peps/pep-0495/#temporal-arithmetic-and-comparison-operators):

> The value of fold will also be ignored whenever a timedelta is added to or subtracted from a datetime instance which may be either aware or naive. The result of addition (subtraction) of a timedelta to (from) a datetime will always have fold set to 0 even if the original datetime instance had fold=1 . 

`pytz` does its own thing and you shouldn't be doing arithmetic on datetimes with `pytz` zones without re-normalizing them afterwards anyway (like for example, consider `d1.astimezone(pytz.timezone('US/Eastern')) - timedelta(hours=2)`).

If you want to add absolute *time* to a datetime (like, say, 5 seconds), it's always best to work in a fixed-offset zone anyway.

**Comment by bdkearns** [comment]
Ah okay. Arithmetic with timedelta may be to spec, but equality does seem flawed. Shouldn't `d2 == d3` return `False`?
> Whenever one or both of the operands in inter-zone comparison is such that its utcoffset() depends on the value of its fold fold attribute, the result is False .

**Comment by pganssle** [comment]
@bdkearns That particular passage doesn't apply, it's not an inter-zone comparison, because both both `d2` and `d3` are in the same zone (`d2.tzinfo is d3.tzinfo == True`).

I'll be honest, I'll have to refresh myself on how exactly the equality works for aware datetimes, but my guess is that this counter-intuitive behavior is more likely a result of Python's datetime equality semantics than `dateutil`'s implementation, since `dateutil` isn't really involved in the equality comparison and returns the right value for `utcoffset()`.

Not actually related, but tangentially related is [this python bug](https://bugs.python.org/issue28601).

**Comment by bdkearns** [comment]
Thanks for the useful links and apologies for the winding report, but what about d1 == d2? It is inter-zone, and value of d2 in UTC depends on fold, so should compare False?

**Comment by pganssle** [comment]
@bdkearns Yeah, that actually is a difference between `_DatetimeWithFold` and Python 3.6's implementation of `datetime`.

@abalkin Maybe you can chime in here? The `_DatetimeWithFold` thing came from your code, basically. What is your recommendation, backport the equality semantics from Python 3.6?

**Comment by pganssle** [comment]
I changed the title to reflect what seems to be the bug or API issue that actually remains after the discussion.

**Comment by abalkin** [comment]
I am not sure what exactly the issue is here.  It is a feature of PEP 495 that two datetime instances that compare equal may produce different timestamps.  Simplifying the original example,

```python
>>> t = datetime.datetime(2016, 11, 6, 1, 12, 15)
>>> t == t.replace(fold=1)
True
>>> t.timestamp() == t.replace(fold=1).timestamp()
False
```
(The system timezone in America/New_York.)

**Comment by pganssle** [comment]
@abalkin The issue is that the `_DatetimeWithFold` implementation is inconsistent between Python 3.5 and 3.6 because it doesn't implement PEP495 equality, so on Python 3.5:

```python
>>> from datetime import datetime, timezone
>>> from dateutil import tz
>>> d1 = datetime(2016, 11, 6, 6, 12, 15, 0, timezone.utc)
>>> d1 == d1.astimezone(tz.gettz('US/Eastern'))
True
```

But in Python 3.6, as demonstrated above, that resolves to `False`.  Since you designed the original `_DatetimeWithFold` implementation, I was wondering if you would have a recommendation on whether I should backport the equality feature into the `_DatetimeWithFold` implementation or if you see another easy way to resolve this to keep the 3.5 -> 3.6 transition backwards compatible.

**Comment by abalkin** [comment]
I see.  It looks like there is an issue with dateutil under Python 3.5 behavior.  Continuing with your session above, if

```python
d2 = d1.astimezone(tz.gettz('US/Eastern'))
```

Under Python 3.6 I get

```python
>>> d1 == d2
False
```

but under Python 3.5

```python
>>> d1 == d2
True
```

Per PEP 495, the `False` result is correct.  This is probably the only ugly feature in the PEP.  See <https://www.python.org/dev/peps/pep-0495/#aware-datetime-equality-comparison>.  As explained in the PEP, times in the fold must always compare as non-equal in order to preserve the hash invariant (x == y implies hash(x) == hash(y)).

I am not sure you can implement PEP 495 behavior in 3.5 by just overriding `__eq__` in `_DatetimeWithFold` because the following is also mandated by PEP 495

```python
>>> d2.replace(fold=0) == d2.replace(fold=0).astimezone(timezone.utc)
False
```

It looks like you will also need to return `_DatetimeWithFold` whenever fold value matters and make `fold` a real member variable rather than an always True property.

**Comment by pganssle** [comment]
@abalkin Good point. I'm not sure it's even possible to do a proper PEP495 implementation before Python 3.6, because even with an implementation of `_DatetimeWithFold` that properly handles the situation you describe, that would *only* work when it was generated by a call to `astimezone`, so we're back to a `pytz`-style `localize`/`normalize` solution, which is not ideal.
